### PR TITLE
revamp prior api

### PR DESCRIPTION
**Pull request by francescoalemanno**
Ty Jerry for all the PRs, and all the work you're putting into this.
@Moelf 
This PR is supposed to do this:
1) Implement the original prior in Scargle's paper + Scargle's correction. i.e. the prior is supposed to be evaluated using the total number of possible splits `N` :
`ncp_prior = 4 - log( fp_rate / ( 0.0136 * length(t) .^ (0.478 ) ) );`
2) <strike>Implement a non-informative prior on the single within-bin counts this is `Jeffrey(0)`, it looks very reasonable on some test data</strike>
implement a easy to use prior like Pearson(p) and Geometric(gamma)
3) Provide an API which is flexible enough to allow users to plug whatever they want.

does this look reasonable?

TODO's:
1) maybe `Scargle(0.05)` is no longer a good default.
2) maybe another kind of prior besides Jeffrey on bin counts must be implemented and offered as default.

### CODE REVIEW

**Comment by Moelf** [comment]
>maybe Scargle(0.05) is no longer a good default.

I think the geometric prior is pretty good, regardless of what the heck data is, `N_b << N` is a good assumption anyway

File: src/BayesHistogram.jl [hunk 1]
```diff
@@ -33,12 +33,33 @@ end
 
 function (w::Jeffrey)(cnt_blocks, cnt_total, cnt_single)
     #                   unused
-    C0 = -0.020833333333333332
-    C1 = -0.730177254404794
-    max(1 + w.penalty, zero(w.penalty)) * log(
-        (cnt_total * sqrt(cnt_total) / 2) /
-        (sqrt(cnt_single) * (C0 + cnt_total * (1 / 4 + C1 * sqrt(cnt_total) + cnt_total))),
-    )
+    ### MATHEMATICA CODE
```
Comment (Moelf) [inline src/BayesHistogram.jl:36]: big brain moment

**Comment by francescoalemanno** [comment]
I added you as a collaborator btw :)

File: src/BayesHistogram.jl (hunk 1 above)
Comment (francescoalemanno) [reply src/BayesHistogram.jl:36]: 🤣

**Comment by Moelf** [comment]
oh yeah we can totally just do this:
image

is this what we need?
```diff
diff --git a/src/BayesHistogram.jl b/src/BayesHistogram.jl
index 1152ef8..d85314e 100644
--- a/src/BayesHistogram.jl
+++ b/src/BayesHistogram.jl
@@ -41,6 +41,16 @@ function (w::Scargle)(n, n_max)
     log(C0*w.p0*n^C1) - 4.0
 end

+struct Geometric{T<:Real, W<:AbstractFloat}
+    gamma::T
+end
+
+function (w::Geometric)(n, n_max)
+    gamma = w.gamma
+    N_p1 = n_max+1
+    (1-gamma)/(1-gamma^N_p1) * gamma^(n)
+end
+
```

**Comment by Moelf** [comment]
oh wait:
```> But Eq. (3) is very convenient
to implement, since with the fitness equal to the log of the posterior,
one only needs to subtract the constant −log γ (called ncp_prior
in the MatLab code

**Comment by francescoalemanno** [comment]
True, I also at some point realised that there is no need for normalisation constants :)
btw I removed that "Jeffrey" monstrosity and added a simple Pearson Chi2 prior and the Geometric prior

**Comment by Moelf** [comment]
image

you didn't update plots it seems

**Comment by francescoalemanno** [comment]
> image
> 
> you didn't update plots it seems

now I fixed it, at some point the geometric distribution was defined without a minus so the gamma had to be small, now it has to be large 👍🏼

**Comment by Moelf** [comment]
hmmm, is there a way to normalize this so gamma can just be from 0 to 1? or no

**Review by Moelf — APPROVED** [review]

**Comment by Moelf** [comment]
wait yeah that looks great and we can use the inner constructor to limit gamma to be 0-1

**Comment by Moelf** [comment]
ok I'm really happy with what we have now feel free to merge

**Pull request merged**